src/compute/src/compute_state.rs

```diff
@@ -937,6 +964,42 @@ impl<'a, A: Allocate + 'static> ActiveComputeState<'a, A> {
         // responses. This happens during shutdown.
         let _ = self.response_tx.send(response);
     }
+
+    /// Returns the dataflow expiration, i.e, the timestamp beyond which diffs can be
+    /// dropped.
+    ///
+    /// Returns an empty timestamp if `replica_expiration` is unset or matches conditions under
+    /// which dataflow expiration should be disabled.
+    pub fn determine_dataflow_expiration<P, S>(
+        &self,
+        plan: &DataflowDescription<P, S, mz_repr::Timestamp>,
+    ) -> Antichain<mz_repr::Timestamp> {
```

Comment:
I agree with this take! Seems like some surrounding code would become less awkward if treated expiry times as optional timestamps instead of frontiers.